server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java

```diff
@@ -49,21 +50,56 @@ protected boolean searchWithCollector(
         boolean hasFilterCollector,
         boolean hasTimeout
     ) throws IOException {
-        return searchWithCollectorManager(searchContext, searcher, query, collectors, hasFilterCollector, hasTimeout);
+        // create the top docs collector
+        final TopDocsCollectorContext topDocsFactory = createTopDocsCollectorContext(searchContext, hasFilterCollector);
+        return searchWithCollector(
+            searchContext,
+            searcher,
+            query,
+            collectors,
+            topDocsFactory,
+            hasFilterCollector,
+            hasTimeout,
+            topDocsFactory.shouldRescore()
```

Comment:
This will not work,  ideally the `shouldRescore` should be invoked **after** the search finishes since it may use whatever state it may have accumulated

Reply:
got it, adding `RescoringQueryCollectorContext` interface should solve this

Reply:
Added `RescoringQueryCollectorContext` interface